### PR TITLE
Adding automatic mixed precision training (via graph rewrite) support to NCF

### PR DESCRIPTION
**Pull request by tgrel**
In continuation of https://github.com/tensorflow/models/pull/7419 and https://github.com/tensorflow/models/pull/7531 this PR adds automatic mixed precision functionality to the NCF model. It can be enabled by running with `--dtype=fp16 --fp16_implementation=graph_rewrite`

### CODE REVIEW

**Comment by tgrel** [comment]
@googlebot I signed it!

**Comment by tgrel** [comment]
@googlebot I consent.

**Comment by vinhngx** [comment]
@googlebot I consent

**Comment by tgrel** [comment]
Here's a test table of the benchmarks I've added for reference:

  | FP32 throughput | AMP throughput | AMP speedup | FP32 hit rate @ 10 | AMP hit rate @ 10
-- | -- | -- | -- | -- | --
1 GPU | 3.20E+06 | 4.38E+06 | 1.37 | 0.6317 | 0.6352
1 GPU + XLA | 2.52E+06 | 5.91E+06 | 2.34 | 0.6295 | 0.6331
8 GPU | 1.33E+06 | 1.02E+07 | 7.70 | 0.6291 | 0.6335
8 GPU (repeated) | 1.33E+06 | 1.02E+07 | 7.69 | 0.6345 | 0.6359

**Comment by saberkun** [comment]
I don't know why cla/google bot still complains. Please double check the agreement is signed.

Thank you for these changes.
I feel we should have a central place to handle optimizer changes like this. Will consider refactoring.

File: official/recommendation/neumf_model.py [hunk 1]
```diff
@@ -118,6 +118,9 @@ def neumf_model_fn(features, labels, mode, params):
       # TODO(seemuch): remove this contrib import
       optimizer = tf.contrib.tpu.CrossShardOptimizer(optimizer)
 
+    if params['fp16_implementation'] == "graph_rewrite":
+      optimizer = tf.train.experimental.enable_mixed_precision_graph_rewrite(optimizer)
```
Comment (saberkun) [inline official/recommendation/neumf_model.py]: Drop the changes in this file?

**Comment by tgrel** [comment]
It looks like someone already made most of the required changes here: https://github.com/tensorflow/models/pull/7436. I've checked and it works fine for TF 2.0. I'm closing this PR and will submit a separate one just for adding some additional benchmarks.

**Pull request closed without merging**